include/circt/Dialect/FIRRTL/SinkLocation.h

```diff
@@ -0,0 +1,108 @@
+//===- SinkLocation.h - FIRRTL Sink Locations -------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This header file defines SinkLocation and helpers for them.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_FIRRTL_SINKLOCATION_H
+#define CIRCT_DIALECT_FIRRTL_SINKLOCATION_H
+
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/Value.h"
+#include "llvm/ADT/DenseMapInfo.h"
+#include "llvm/ADT/SmallVector.h"
+
+namespace circt {
+namespace firrtl {
+
+/// This class represents a reference to a specific field or element of a sink
+/// value. This is used to distinguish between the individual elements of Values
+/// which are aggregate sink types.  This is capable of referring to a field in
+/// a bundle, or a element in a vector. All indexing operations referring to the
+/// same field in an aggregate have the same SinkLocation.  This can be used as
+/// the key in a hashtable to store field specific information. It is not
+/// possible to go from a SinkLocation back to the Subfield op it was created
+/// from.
+class SinkLocation {
+public:
+  /// Get a null SinkLocation.
+  SinkLocation(){};
+
+  /// Get a sink location for the specified value.  If the value is the result
+  /// of indexing in to an aggregate type, this will resolve the location to be
+  /// a member of that aggregate.
+  SinkLocation(Value value);
+
+  /// Get the Value which created this sink location.
+  Value getSink() const { return sink; }
+
+  /// Get the operation which defines this sink.
+  Operation *getDefiningOp() const;
+
+  /// Get a string representation of this sink location.
+  std::string getFieldName() const;
+
+  /// Get the path of indices to this specific field.
+  std::vector<unsigned> &getPath() { return path; }
+  const std::vector<unsigned> &getPath() const { return path; }
+
+  bool operator==(const SinkLocation &other) const {
+    if (sink != other.sink)
+      return false;
+    return std::equal(path.begin(), path.end(), other.path.begin(),
+                      other.path.end());
+  }
+
+  /// Print the location.
+  void print(raw_ostream &os) const;
+  void dump() const;
+
+private:
+  /// A pointer to the value which created this sink location.
+  Value sink;
+
+  /// The path through bundle and array accesses that this sink represents.
+  std::vector<unsigned> path;
```

Comment:
This could be a single integer representing the field index from a depth-first walk or something.

Reply:
One thing that isn't clear to me is whether or not we should be able to create locations for non-leaf elements.   If only leaf-elements are assigned IDs, they could be reusable as an array index if someone wanted to represent leaf-only data using an array instead of a hashtable. OTOH allowing a location to be constructed for any field seems more consistent to me.

Reply:
I went with allowing refs to any field.